apic_ml2/neutron/db/l3out_vlan_allocation.py

```diff
@@ -0,0 +1,241 @@
+# Copyright (c) 2013 OpenStack Foundation
+# All Rights Reserved.
+#
+#    Licensed under the Apache License, Version 2.0 (the "License"); you may
+#    not use this file except in compliance with the License. You may obtain
+#    a copy of the License at
+#
+#         http://www.apache.org/licenses/LICENSE-2.0
+#
+#    Unless required by applicable law or agreed to in writing, software
+#    distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+#    WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+#    License for the specific language governing permissions and limitations
+#    under the License.
+
+from oslo_log import log
+from six import moves
+import sqlalchemy as sa
+
+from neutron.common import exceptions as exc
+from neutron.db import api as db_api
+from neutron.db import model_base
+from neutron.i18n import _LE, _LI, _LW
+from neutron.plugins.common import constants as p_const
+from neutron.plugins.common import utils as plugin_utils
+from neutron.plugins.ml2.drivers import helpers
+
+LOG = log.getLogger(__name__)
+
+
+class L3OutVlanAllocation(model_base.BASEV2):
+    """Represent allocation state of a vlan_id for the L3 out per VRF.
+
+    If allocated is False, the vlan_id is available for allocation.
+    If allocated is True, the vlan_id is in use.
+
+    When an allocation is released, if the vlan_id is inside the pool
+    described by network_vlan_ranges, then allocated is set to
+    False. If it is outside the pool, the record is deleted.
+    """
+
+    __tablename__ = 'apic_ml2_l3out_vlan_allocation'
+    __table_args__ = (
+        sa.Index('apic_ml2_l3out_vlan_allocation_l3out_network_allocated',
+                 'l3out_network', 'allocated'),
+        model_base.BASEV2.__table_args__,)
+
+    l3out_network = sa.Column(sa.String(64), nullable=False,
+                              primary_key=True)
+    vrf = sa.Column(sa.String(64), nullable=False,
+                    primary_key=False)
+    vlan_id = sa.Column(sa.Integer, nullable=False, primary_key=True,
+                        autoincrement=False)
+    allocated = sa.Column(sa.Boolean, nullable=False)
+
+
+class NoVlanAvailable(exc.ResourceExhausted):
+    message = _("Unable to allocate the vlan. "
+                "No vlan is available for %(l3out_network)s external network")
+
+
+# inherit from SegmentTypeDriver to reuse the code to reserve/release
+# vlan IDs from the pool
+class L3outVlanAlloc(helpers.SegmentTypeDriver):
+
+    def __init__(self):
+        super(L3outVlanAlloc, self).__init__(L3OutVlanAllocation)
+        self.session = db_api.get_session()
+
+    def _parse_vlan_ranges(self, ext_net_dict):
+        self.l3out_vlan_ranges = {}
+        try:
+            for l3out_network in ext_net_dict.keys():
+                ext_info = ext_net_dict.get(l3out_network)
+                vlan_range_str = ext_info.get('shadow_l3out_vlan_range')
```

Comment:
'shadow' in the option name doesn't make sense (from the perspective of end user), and 'l3out' is redundant. I'd suggest you name this option as "vlan_range"